JBPM-7860: case modeler - support the new implementation of marshaller

xiezhang7 opened the pull request.

xiezhang7 commented:
Hi @romartin @hasys @LuboTerifaj @evacchi @Josephblt ,

This is the pull request to support the new marshaller for case modeler.
[JBPM-7981](https://issues.jboss.org/browse/JBPM-7981) is also fixed by this pull request.

Thanks.

kiereleaseuser3 commented:
Can one of the admins verify this PR? Comment with 'ok to test' to start the build.

kiereleaseuser2 commented:
Can one of the admins verify this PR? Comment with 'ok to test' to start the build.

kiereleaseuser2 commented:
Can one of the admins verify this PR? Comment with 'ok to test' to start the build.

kiereleaseuser2 commented:
Can one of the admins verify this PR? Comment with 'ok to test' to start the build.

kiereleaseuser2 commented:
Can one of the admins verify this PR? Comment with 'ok to test' to start the build.

romartin commented:
ok to test

Josephblt commented:
Hey @xiezhang7. Did a quick test. The CM saved and loaded just fine. Also inspected the file it seems to be working great. Looks good to me.

xiezhang7 commented:
Thanks @Josephblt .

LuboTerifaj commented:
Jenkins execute full downstream build

xiezhang7 commented:
Jenkins execute full downstream build

xiezhang7 commented:
Hi @evacchi ,

Thanks for the feedback.

Because the client logic of diagram, adhoc subprocess and reusable subprocess is specialized in case modeler, the following classes were created for the elements of case modeler:
1. [CaseManagementDiagram](https://github.com/kiegroup/kie-wb-common/blob/d629fbad640547234a3f736f36582bb7abc2e48a/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-api/src/main/java/org/kie/workbench/common/stunner/cm/definition/CaseManagementDiagram.java#L63) : equivalent to `BPMNDiagramImpl` in BPMN
2. [AdHocSubprocess](https://github.com/kiegroup/kie-wb-common/blob/d629fbad640547234a3f736f36582bb7abc2e48a/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-api/src/main/java/org/kie/workbench/common/stunner/cm/definition/AdHocSubprocess.java#L60) : equivalent to `AdHocSubprocess` in BPMN
3. [CaseReusableSubprocess](https://github.com/kiegroup/kie-wb-common/blob/d629fbad640547234a3f736f36582bb7abc2e48a/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-api/src/main/java/org/kie/workbench/common/stunner/cm/definition/CaseReusableSubprocess.java#L51) : equivalent to `ReusableSubprocess` in BPMN
4. [ProcessReusableSubprocess](https://github.com/kiegroup/kie-wb-common/blob/d629fbad640547234a3f736f36582bb7abc2e48a/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-api/src/main/java/org/kie/workbench/common/stunner/cm/definition/ProcessReusableSubprocess.java#L50) : equivalent to `ReusableSubprocess` in BPMN
5. [CaseReusableSubprocessTaskExecutionSet](https://github.com/kiegroup/kie-wb-common/blob/d629fbad640547234a3f736f36582bb7abc2e48a/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-api/src/main/java/org/kie/workbench/common/stunner/cm/definition/property/task/CaseReusableSubprocessTaskExecutionSet.java#L48) : equivalent to `ReusableSubprocessTaskExecutionSet` in BPMN
6. [ProcessReusableSubprocessTaskExecutionSet](https://github.com/kiegroup/kie-wb-common/blob/d629fbad640547234a3f736f36582bb7abc2e48a/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-api/src/main/java/org/kie/workbench/common/stunner/cm/definition/property/task/ProcessReusableSubprocessTaskExecutionSet.java#L48) : equivalent to `ReusableSubprocessTaskExecutionSet` in BPMN

Marshalling logic for these CM elements is supposed to be the same as the marshalling for the BPMN elements.
I changed the related CM and BPMN elements to inherit the same abstract super class or interface, so that marshalling code can avoid relying on BPMN logic and be shared for both BPMN and CM:
1. [BPMNDiagram](https://github.com/kiegroup/kie-wb-common/blob/d629fbad640547234a3f736f36582bb7abc2e48a/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/main/java/org/kie/workbench/common/stunner/bpmn/definition/BPMNDiagram.java#L21) for `BPMNDiagramImpl` and `CaseManagementDiagram`
2. [BaseAdHocSubprocess](https://github.com/kiegroup/kie-wb-common/blob/4a9ea9a7ff3c20d5f634daf468ead951052e0339/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/main/java/org/kie/workbench/common/stunner/bpmn/definition/BaseAdHocSubprocess.java#L26) for BPMN `AdHocSubprocess` and CM `AdHocSubprocess`
3. [BaseReusableSubprocess](https://github.com/kiegroup/kie-wb-common/blob/4a9ea9a7ff3c20d5f634daf468ead951052e0339/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/main/java/org/kie/workbench/common/stunner/bpmn/definition/BaseReusableSubprocess.java#L26) for `ReusableSubprocess` `CaseReusableSubprocess` and `ProcessReusableSubprocess`
4. [BaseReusableSubprocessTaskExecutionSet](https://github.com/kiegroup/kie-wb-common/blob/4a9ea9a7ff3c20d5f634daf468ead951052e0339/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/main/java/org/kie/workbench/common/stunner/bpmn/definition/property/task/BaseReusableSubprocessTaskExecutionSet.java#L20) for `ReusableSubprocessTaskExecutionSet` `CaseReusableSubprocessTaskExecutionSet` and `ProcessReusableSubprocessTaskExecutionSet`

Some new marshaller code is changed to used these abstract classes, instead of having dependencies on BPMN code directly:
1. Created `BaseDirectDiagramMarshaller`, `BaseConverterFactory` (from), `BaseFlowElementConverter`, `BaseReusableSubprocessConverter`, `BaseSubProcessConverter` (from), `BaseConverterFactory` (to), `BaseRootProcessConverter`, `BaseSubProcessConverter` (to) and `BaseCallActivityConverter` for the shared code.
2. Kept BPMN dependencies in `BPMNDirectDiagramMarshaller`, `ConverterFactory` (from), `FlowElementConverter`, `ReusableSubprocessConverter`, `SubProcessConverter` (from), `ConverterFactory` (to), `CallActivityConverter`, `RootProcessConverter`, `SubProcessConverter` (to).
3. Modified classes `DefinitionsConverter`, `EdgeConverter`, `ProcessConverterDelegate`, `RootProcessConverter`, `FlowElementConverter` and `ProcessConverterDelegate` to use the abstract classes.

Most of the changes to the new marshaller code are refactoring changes. The only thing new that I added is a new constructor to [DefinitionsBuildingContext](https://github.com/kiegroup/kie-wb-common/blob/4a9ea9a7ff3c20d5f634daf468ead951052e0339/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-backend/src/main/java/org/kie/workbench/common/stunner/bpmn/backend/converters/fromstunner/DefinitionsBuildingContext.java#L66). Because I need to use `CaseManagementDiagram` instead of `BPMNDiagramImpl` to create instance of `DefinitionsBuildingContext`.

Regarding the docs, I just checked them. I think everything in the docs is still valid, and there is nothing really new to be add to the docs. 

Thanks.

xiezhang7 commented:
Jenkins execute full downstream build

LuboTerifaj commented:
Jenkins execute full downstream build

wmedvede commented:
Hi, @xiezhang7  @Josephblt , which showcase do you use guys for testing this CM features?

xiezhang7 commented:
Hi @wmedvede , I tested it with the standalone showcase and business central webapp.

xiezhang7 commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-backend/src/main/java/org/kie/workbench/common/stunner/bpmn/backend/BaseDirectDiagramMarshaller.java:
```diff
@@ -0,0 +1,258 @@
+import org.slf4j.LoggerFactory;
+
+/*
+ * Direct as in "skipping json encoding"
```
Hi @LuboTerifaj , I am also not sure what this means.
It is not new code. It is in [BPMNDirectDiagramMarshaller](https://github.com/kiegroup/kie-wb-common/blob/0007ba94bd8d3033ac404b5b94119cbcb7acca92/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-backend/src/main/java/org/kie/workbench/common/stunner/bpmn/backend/BPMNDirectDiagramMarshaller.java#L73) . I only extracted the code to this new abstract super class.

evacchi replied:
I can answer that: the name BPMN**Direct**DiagramMarshaller was opposed to the existing BPMNDiagramMarshaller. It's not a good name, I couldn't come up with a better one at the time. The old BPMNDiagramMarshaller went through an additional step converting XML into JSON for legacy reasons. The reason for the new version, beside a necessary spring cleaning, was to remove this extra step. So the new version is "Direct" as in "it doesn't go through the extra step". Another provisional name was BPMNJsonLessDiagramMarshaller, which is even worse. 😅 Anyway the comment is there just to justify the bad naming choice.

xiezhang7 replied:
Hi @LuboTerifaj @evacchi , I have put the information to the comments.

xiezhang7 commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-api/src/main/java/org/kie/workbench/common/stunner/cm/definition/property/task/BaseCaseManagementReusableSubprocessTaskExecutionSet.java:
```diff
@@ -0,0 +1,27 @@
+public interface BaseCaseManagementReusableSubprocessTaskExecutionSet
+        extends BaseReusableSubprocessTaskExecutionSet {
+
+    Case getCaze();
```
Hi @LuboTerifaj , this is because the field name is "caze" in the subclasses [CaseReusableSubprocessTaskExecutionSet](https://github.com/kiegroup/kie-wb-common/blob/dcee3e9d1a685c6a2f69938be7f36ec1bee22d89/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-api/src/main/java/org/kie/workbench/common/stunner/cm/definition/property/task/CaseReusableSubprocessTaskExecutionSet.java#L66) and [ProcessReusableSubprocessTaskExecutionSet](https://github.com/kiegroup/kie-wb-common/blob/dcee3e9d1a685c6a2f69938be7f36ec1bee22d89/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-api/src/main/java/org/kie/workbench/common/stunner/cm/definition/property/task/ProcessReusableSubprocessTaskExecutionSet.java#L66) . "case" is a keyword in java, I cannot use it as field name. So I used "caze" instead.

evacchi replied:
my taste: I'd still call the getter/setters `getCase`.

xiezhang7 replied:
Hi @LuboTerifaj @evacchi , I changed the name to "IsCase".

xiezhang7 commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-backend/src/main/java/org/kie/workbench/common/stunner/bpmn/backend/BaseDirectDiagramMarshaller.java:
```diff
@@ -0,0 +1,258 @@
+        DefinitionsConverter definitionsConverter =
+                new DefinitionsConverter(createFromStunnerConverterFactory(diagram.getGraph(),
+                                                                           propertyWriterFactory),
+                                         propertyWriterFactory);
```
Hi @LuboTerifaj , I have fixed the formatting.

xiezhang7 commented:
Jenkins execute full downstream build

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-backend/src/main/java/org/kie/workbench/common/stunner/bpmn/backend/BaseBackendService.java:
```diff
@@ -0,0 +1,44 @@
+import org.kie.workbench.common.stunner.core.graph.Graph;
+import org.slf4j.LoggerFactory;
+
+public abstract class BaseBackendService extends AbstractDefinitionSetService {
```
This class does not do too much... right? no abstract fields, no class members, just inherits and nothing special, as the `chooseMarshallers` is also static... is it really necessary?

xiezhang7 replied:
Hi @romartin , I have removed this class.

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-backend/src/main/java/org/kie/workbench/common/stunner/cm/backend/CaseManagementDirectDiagramMarshaller.java:
```diff
@@ -0,0 +1,295 @@
+    private void preMarshallProcess(final Diagram<Graph, Metadata> diagram) {
+        // adjust the position of the elements
+        StreamSupport.stream(diagram.getGraph().nodes().spliterator(), false)
+                .filter(node -> !(((View) ((NodeImpl) node).getContent()).getDefinition() instanceof CaseManagementDiagram))
```
This filter operation looks a bit confusing due to casts and the use of the `instanceof`... I would suggest first creating a new variable that has the right type and also use the method `Class#isInstance` as well, something like: 
```
Iterable<Node<View<?>, Edge>> nodesIt = diagram.getGraph().nodes();
StreamSupport.stream(nodesIt.spliterator(), false)
    .filter(node -> CaseManagementDiagram.class.isInstance(node.getContent().getDefinition()))
   ...
```
It looks more clear, isn't it? Well it's just a suggestion, the code is working fine anyway, but readability helps... :)

xiezhang7 replied:
Thanks @romartin , I have changed the code as you suggested.

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-backend/src/main/java/org/kie/workbench/common/stunner/cm/backend/CaseManagementDirectDiagramMarshaller.java:
```diff
@@ -0,0 +1,295 @@
+                });
+
+        StreamSupport.stream(diagram.getGraph().nodes().spliterator(), false)
+                .filter(node -> ((View) ((NodeImpl) node).getContent()).getDefinition() instanceof CaseManagementDiagram)
```
same as above comment here

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-backend/src/main/java/org/kie/workbench/common/stunner/cm/backend/CaseManagementDirectDiagramMarshaller.java:
```diff
@@ -0,0 +1,295 @@
+    @SuppressWarnings("unchecked")
+    private void postUnmarshallProcess(final Graph<DefinitionSet, Node> graph) {
+        StreamSupport.stream(graph.nodes().spliterator(), false)
+                .filter(node -> ((View) node.getContent()).getDefinition() instanceof CaseManagementDiagram)
```
same here

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-backend/src/main/java/org/kie/workbench/common/stunner/cm/backend/CaseManagementDirectDiagramMarshaller.java:
```diff
@@ -0,0 +1,295 @@
+
+            // remove sequence flow between elements in stages
+            root.getOutEdges().stream().map(e -> ((Edge) e).getTargetNode())
+                    .filter(n -> ((View) ((Node) n).getContent()).getDefinition() instanceof AdHocSubprocess)
```
same here

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-backend/src/main/java/org/kie/workbench/common/stunner/cm/backend/CaseManagementDirectDiagramMarshaller.java:
```diff
@@ -0,0 +1,295 @@
+
+        // reset position of elements
+        StreamSupport.stream(graph.nodes().spliterator(), false)
+                .filter(node -> !(((View) node.getContent()).getDefinition() instanceof CaseManagementDiagram))
```
same here

Josephblt commented:
Hey Walter, sorry for the delay. I'm using Stunner Product Showcase. There
is a trick to make it work. For the project to be recognized as a Case
Project there is a trick.
You have to put a file with the extension .caseProject on the root folder
of your project.

Create a new regular project. There is only one option for project creation.
Clone the repository for your project: git clone .niogit/MySpace/[Name of
your project].git.
Than create a file named .caseProject, commit and push.

That's it. The project will be recognized as a case project and Case
Definition Designer will be available.
For normal BPMN Designer just check as adhoc and CM properties will appear.
For case BPMN Desiner (Case Definition) the properties appear by default,
if I recall correctly.

If anything just ping me on IRC (scholl).
Again sorry for the delayed response. This email went unchecked yesterday,
I closed it and just noticed now.


Em ter, 11 de dez de 2018 às 14:09, Walter Medvedeo <
notifications@github.com> escreveu:

> Hi, @xiezhang7 <https://github.com/xiezhang7> @Josephblt
> <https://github.com/Josephblt> , which showcase do you use guys for
> testing this CM features?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kiegroup/kie-wb-common/pull/2328#issuecomment-446259402>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AYMakmrV9MDOiEOHn_GUK24DSUPhg2N1ks5u39i3gaJpZM4ZBGlc>
> .
>

wmedvede commented:
> Hey Walter, sorry for the delay. I'm using Stunner Product Showcase. There is a trick to make it work. For the project to be recognized as a Case Project there is a trick. You have to put a file with the extension .caseProject on the root folder of your project. Create a new regular project. There is only one option for project creation. Clone the repository for your project: git clone .niogit/MySpace/[Name of your project].git. Than create a file named .caseProject, commit and push. That's it. The project will be recognized as a case project and Case Definition Designer will be available. For normal BPMN Designer just check as adhoc and CM properties will appear. For case BPMN Desiner (Case Definition) the properties appear by default, if I recall correctly. If anything just ping me on IRC (scholl). Again sorry for the delayed response. This email went unchecked yesterday, I closed it and just noticed now. Em ter, 11 de dez de 2018 às 14:09, Walter Medvedeo < notifications@github.com> escreveu:
> […](#)
> Hi, @xiezhang7 <https://github.com/xiezhang7> @Josephblt <https://github.com/Josephblt> , which showcase do you use guys for testing this CM features? — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#2328 (comment)](https://github.com/kiegroup/kie-wb-common/pull/2328#issuecomment-446259402)>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AYMakmrV9MDOiEOHn_GUK24DSUPhg2N1ks5u39i3gaJpZM4ZBGlc> .

Good trick :), thanks Wagner.

LuboTerifaj commented:
Jenkins execute full downstream build

wmedvede commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-backend/src/test/java/org/kie/workbench/common/stunner/cm/backend/CaseManagementDirectDiagramMarshallerTest.java:
```diff
@@ -0,0 +1,440 @@
+        process2Node.getInEdges().add(process2InEdge);
+
+        DefinitionSet definitionSet = new DefinitionSetImpl("org.kie.workbench.common.stunner.cm.CaseManagementDefinitionSet");
+        definitionSet.setBounds(BoundsImpl.build(0.0, 0.0, 2800.0, 1400.0));
```
Hi @xiezhang7 just for the record, but this line produced compile errors in my local when I rebased and built my stuff, with this PR's stuff, plus master leatest stuff this morning.
It seems like the DefininitionSet was changed by @romartin Yesterday commit https://github.com/kiegroup/kie-wb-common/commit/1082e1994c9ed231384984b8bb8b9da2205b56fe and the setBounds(XXX) method is no longer available. I believe it might be worth rebasing this PR with master and check if it's building well. And had also update and build locally the lienzo-core project for making everything build well.
Regards,
Walter.

xiezhang7 replied:
Hi @wmedvede , thanks a lot! I have rebased this PR to master.

xiezhang7 commented:
Hi @romartin , I have added more tests to the new marshaller in [CaseManagementDirectDiagramMarshallerTest](https://github.com/kiegroup/kie-wb-common/blob/7a263b142d5b8a1d3f35312bc74e012cc62c61f1/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-case-mgmt/kie-wb-common-stunner-case-mgmt-backend/src/test/java/org/kie/workbench/common/stunner/cm/backend/CaseManagementDirectDiagramMarshallerTest.java#L134) .

xiezhang7 commented:
Jenkins execute full downstream build

xiezhang7 commented:
Jenkins execute full downstream build

xiezhang7 commented:
Hi @romartin @LuboTerifaj , 

I don't think the build [1180](https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/KIE/job/master/job/pullrequest/job/kie-wb-common-pullrequests/1180/) failed because of code changes in this pull request. Because the same test `EmbeddedControllerIT` also failed for build [1177](https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/KIE/job/master/job/pullrequest/job/kie-wb-common-pullrequests/1177/) with the same error. I don't think code change in this pull request has anything to do with that test. 

And because of some maven dependency issues, I could not run the failing test locally on my machine. So I'll try to re-build the pull request instead and see if the test will pass this time.

Thanks.

xiezhang7 commented:
Jenkins execute full downstream build

LuboTerifaj commented:
There is still missing full downstream build.

LuboTerifaj commented:
Jenkins execute full downstream build

xiezhang7 commented:
Thanks @LuboTerifaj !

xiezhang7 commented:
Hi @romartin , could you please merge this pull request? Thanks.
